exonum-java-binding/core/rust/src/storage/db.rs

```diff
@@ -28,104 +28,120 @@ pub(crate) type Value = Vec<u8>;
 ///    in the case when the java side creates and owns it;
 /// - it just holds a reference, when one is provided from the rust side.
 ///
-/// For storage API we need a reference, so we create it to the owned part. But since there is no
-/// way in Rust to make a `View` value not movable. Furthermore, it have to be moved from the stack
-/// to the heap in order to be converted into `Handle` for the java side. So a `Fork` value
-/// should be placed in the heap to prevent its movement after creating a reference to it.
-
-pub(crate) struct View {
-    /// Unsafe reference to the owned View (owned either by Rust or by Java)
-    reference: ViewRef,
-    /// Allows to drop owned view if needed
-    owned: Option<ViewOwned>,
+/// As `View` does not have a lifetime, nothing protects us from dereferencing a freed reference,
+/// the extreme caution must be taken in places where `View` is constructed and send to Java.
+/// Java code must never store a `View` handle beyond the scope it was initially acquired
+/// (except for `View::Owned`).
+pub(crate) enum View {
+    /// Immutable Fork view, constructed from `&Fork`.
+    ///
+    /// Created `View` must never outlive the reference it was created with,
+    /// or `SIGINT` will occur.
+    RefFork(&'static Fork),
+    /// Mutable Fork view, constructed from `&mut Fork`.
+    ///
+    /// Created `View` must never outlive the reference it was created with,
+    /// or `SIGINT` will occur.
+    RefMutFork(&'static mut Fork),
+    /// Immutable Snapshot view, constructed from `&Snapshot`. There is no need in mutable
+    /// variant.
+    ///
+    /// Created `View` must never outlive the reference it was constructed with,
+    /// or `SIGINT` will occur.
+    RefSnapshot(&'static dyn Snapshot),
+    /// Covers both `Snapshot` and `Fork` cases. Rust uses move semantic and single-ownership
```

Comment:
I'd clarify, what Owned means. If I understand correctly, the
client code owns a Snapshot/View object and then transfers the ownership of it to the View, which then owns it. I.e. 'owned' means different things in different contexts (= when used in factory method below and here in enum variant).

Reply:
You're right, but I can't see why the meaning in different here and there. 🤔 We have "owned" Snaphot/Fork -> we use `from_owned` method -> we have "owned" View, i.e. "View that owns Snapshot/Fork", which is the opposite of "View that borrows Snapshot/Fork".

Reply:
I mean that the subject which owns the snapshot/fork is different, but that it probably isn't that important.